### DISCUSSION TITLE
How to convert HEIC to jpeg with -strip with minimal color shifts?

### DISCUSSION THREAD

**Discussion by davidsalazar** [ImageMagick/ImageMagick#4391] [Help] [answered]
So taking an iPhone heic photo and just converting to jpeg and stripping exif will screw up all the colors due to the non-srgb color profiles the iPhone uses. 

Is there a way to detect whether a non-srgb profile is used and convert it to SRGB instead of having to specify both the source and destination icc profiles?

Looking at this [thread](https://github.com/ImageMagick/ImageMagick/issues/1486), I'm currently doing the following with pretty good results but as you can see, it's not ideal having to specify the source ICC profile. Seems like this should be automated since the source icc profile is already embedded in the source image. Do we really have to write extra logic to extract it using a tool, examine whether it should be converted to SRGB or not? Surely this is a super common task isn't it? 

`magick source.HEIC -colorspace sRGB -profile P3.icc -profile c2.icc -strip -auto-orient -interlace Plane -gaussian-blur 0.05 -resize 2000x2000 -quality 85% output_should_have_similar_colors.jpg`

**Comment by fmw42** [comment]
If there is a non-sRGB source profile, just add a sRGB profile once. Then strip. You can determine if profiles exist from 

`magick image.heic -format "%[profiles]" info:`

and/or

```
magick image.heic -format "%[profile:icc]" info:
magick image.heic -format "%[profile:icm]" info:

```

If there is no profile, then you would have to include an initial appropriate profile.

**Reply by davidsalazar** [reply]
Thanks for the reply. When you say just add the SRGB profile once, I just tried tried that now. I took the command above and removed the source profile and doing that screwed up colors. It appears I need that source profile, don't i?

**Reply by davidsalazar** [reply]
Here is some more info if you'd like to replicate the issue. 
[Here is the test package.](https://www.dropbox.com/s/d37w3vnhwxkavl9/ImageMagickTest.zip?dl=0) 

**This Works**
`magick IMG_2253.HEIC -colorspace sRGB -profile P3.icc -profile c2.icc -strip -auto-orient -interlace Plane -gaussian-blur 0.05 -resize 3000x3000 -quality 85% output_should_have_correct_color.jpg`

**This Doesn't Work - colors are skewed.**
`magick IMG_2253.HEIC -colorspace sRGB -profile c2.icc -strip -auto-orient -interlace Plane -gaussian-blur 0.05 -resize 3000x3000 -quality 85% output_should_have_correct_color.jpg`

**Reply by fmw42** [reply]
I did not say remove the source profile. I just said add a srgb profile to convert the source profile to srgb. Then remove the source profile. If you do not have a source profile, then you need to add two profiles.

**Reply by davidsalazar** [reply]
yeah the source profile is embedded but when I just try to convert it to the new source profile, colors get screwed up. When i add the source profile on the cmd line it works fine.

**Reply by snibgo** [reply]
> yeah the source profile is embedded but when I just try to convert it to the new source profile, colors get screwed up.

That is not the case for me. Please show your commands and results. For example:
```
%IMG7%magick IMG_2253.HEIC -profile sRGB.icc x1.jpg

%IMG7%magick IMG_2253.HEIC -profile P3.icc -profile sRGB.icc x2.jpg

%IMG7%magick compare -metric RMSE x1.jpg x2.jpg NULL:
```
The RMSE is "2.40979 (3.67711e-05)", which is virtually no difference.

They both look like this:
![x1](https://user-images.githubusercontent.com/33812368/138190941-3cbd56cb-6ef2-4e10-be20-4406b5009eb8.jpg)

**Reply by davidsalazar** [reply]
The source is p3 and I want the output to be srgb. Looks like you're just converting it to jpeg and embedding the same profile that was in the source. Thats why it looks identical to you.  

If you look the cmds I'm running above, I'm trying to convert the iPhone photo which is in p3 to srgb without drastic color changes.

**Reply by davidsalazar** [reply]
which srgb.icc are you using, I'll run that cmd that you did and I'll show you my results. 

Running this cmd. c2.icc being facebooks srgb profile. Look what I get.
`magick IMG_2253.HEIC -profile c2.icc x1.jpg`

![x1](https://user-images.githubusercontent.com/14309/138191573-edf36295-d348-42c7-8b1c-e4e0c8928a01.jpg)

**Reply by davidsalazar** [reply]
it appears I have to use -colorspace srgb after all. If I leave it out I get the picture I embedded above. When I add colorspace srgb,  I get the same effect as before. Colors slightly skewed when I leave out the source color profile and somewhat acceptable otherwise. 

See 
`$ magick IMG_2253.HEIC -colorspace sRGB -profile c2.icc x1.jpg`

`$ magick IMG_2253.HEIC -colorspace sRGB -profile P3.icc -profile c2.icc x2.jpg`

`$ magick compare -metric RMSE x1.jpg x2.jpg NULL:`
1282.66 (0.0195721)%

**Reply by snibgo** [reply]
Here, I am using sRGB.icc as distributed with ImageMagick.

Something has gone badly wrong to get that result. I don't know what. A developer may need to take a look.

**Reply by davidsalazar** [reply]
Looks like that fixed it. That c2.icc profile I was using was really screwing it up. I tried the one with the imagemagick package and it appears that fixed all my problems lol. How embarrassing. Thanks for all the help. 

But yes, with the srgb.icc from iamgemagick I can now just do
`magick IMG_2253.HEIC -profile srgb.icc -strip x1.jpg` 
AND IT WORKS AS I WANTED. Colors skewed is minimal.

**Comment by snibgo** [comment]
What version of IM, on what platform?

Your "-colorspace sRGB" is redundant.

All HEIC files I have seen contain a "Display P3" ICC profile. If you want a JPG file then:
```
magick IMG_2253.HEIC out.jpg
```
This writes the embedded DisplayP3 profile to out.jpg.

If you want to convert it to a sRGB profile:
```
magick IMG_2253.HEIC -profile sRGB.icc out.jpg
```
... or use your favourite sRGB ICC profile.

You don't need to specify a source profile because that is the embedded profile.

What is your c2.icc? It contains no helpful metadata that says what it is. It seems to be sRGB, but perhaps it isn't.

For accurate colours, I do not recommend stripping ICC profiles. This is because there are many varieties of sRGB, and for accuracy we need to know which was used to encode pixel values.

sRGB has a smaller gamut than Display P3, so colours will not look the same. You may find an "-intent" setting before the "-profile" is helpful.

**Reply by davidsalazar** [reply]
Hi thanks for the reply.
I'm going to try to answer most of your questions in numbers.

1. I'm on ImageMagick 7.0.10-28 on latest macOS. 
2. I realize I can just embed the same color profile to the output jpeg and it does work great, but I was trying to strip it out for web. Google insights doesn't like color profiles. 
3. If I don't specify the color profile, it will drastically change colors when converted. [See this comment where I try it and provide sample package where you can replicate it. ](https://github.com/ImageMagick/ImageMagick/discussions/4391#discussioncomment-1506129)
4. c2.icc is facebooks srgb profile. [See this.](https://photosauce.net/blog/post/making-a-minimal-srgb-icc-profile-part-2-curve-the-curves)
5. I fully understand that srgb has a smaller gamut than p3 and I realize it won't be exactly the same. I also know that converting to srgb doesn't have to look drastically different like it's doing now if I don't include both the source profile in the cmd above when converting to srgb.

**Comment by snibgo** [comment]
Please show the text output from "magick IMG_2253.HEIC -verbose info:". Here is what I get:
```
Image:
  Filename: IMG_2253.HEIC
  Format: HEIC (High Efficiency Image Format)
  Mime type: image/heic
  Class: DirectClass
  Geometry: 3024x4032+0+0
  Resolution: 72x72
  Print size: 42x56
  Units: PixelsPerInch
  Colorspace: sRGB
  Type: TrueColor
  Base type: Undefined
  Endianness: Undefined
  Depth: 8-bit
  Channel depth:
    Red: 8-bit
    Green: 8-bit
    Blue: 8-bit
  Channel statistics:
    Pixels: 12192768
    Red:
      min: 0  (0)
      max: 255 (1)
      mean: 122.21 (0.479257)
      median: 178 (0.698039)
      standard deviation: 63.4709 (0.248906)
      kurtosis: -1.24078
      skewness: -0.259379
      entropy: 0.965867
    Green:
      min: 0  (0)
      max: 255 (1)
      mean: 101.841 (0.399378)
      median: 138 (0.541176)
      standard deviation: 46.4542 (0.182173)
      kurtosis: -0.718845
      skewness: 0.0321177
      entropy: 0.939907
    Blue:
      min: 0  (0)
      max: 255 (1)
      mean: 82.4718 (0.323419)
      median: 90 (0.352941)
      standard deviation: 36.9435 (0.144877)
      kurtosis: 3.22477
      skewness: 1.28199
      entropy: 0.885276
  Image statistics:
    Overall:
      min: 0  (0)
      max: 255 (1)
      mean: 102.175 (0.400684)
      median: 135.333 (0.530719)
      standard deviation: 48.9562 (0.191985)
      kurtosis: -0.690926
      skewness: 0.366626
      entropy: 0.93035
  Rendering intent: Perceptual
  Gamma: 0.454545
  Chromaticity:
    red primary: (0.64,0.33)
    green primary: (0.3,0.6)
    blue primary: (0.15,0.06)
    white point: (0.3127,0.329)
  Matte color: grey74
  Background color: white
  Border color: srgb(223,223,223)
  Transparent color: none
  Interlace: None
  Intensity: Undefined
  Compose: Over
  Page geometry: 3024x4032+0+0
  Dispose: Undefined
  Iterations: 0
  Compression: Undefined
  Orientation: TopLeft
  Profiles:
    Profile-exif: 2520 bytes
    Profile-icc: 548 bytes
  Properties:
    date:create: 2021-10-20T06:26:23+00:00
    date:modify: 2021-10-19T17:43:36+00:00
    exif:ApertureValue: 54823/32325
    exif:BrightnessValue: -8027/2240
    exif:ColorSpace: 65535
    exif:ComponentsConfiguration: 1, 2, 3, 0
    exif:DateTime: 2021:09:17 20:25:29
    exif:DateTimeDigitized: 2021:09:17 20:25:29
    exif:DateTimeOriginal: 2021:09:17 20:25:29
    exif:ExifOffset: 260
    exif:ExifVersion: 48, 50, 51, 50
    exif:ExposureBiasValue: 7/256
    exif:ExposureMode: 0
    exif:ExposureProgram: 2
    exif:ExposureTime: 1/15
    exif:Flash: 16
    exif:FlashPixVersion: 48, 49, 48, 48
    exif:FNumber: 9/5
    exif:FocalLength: 17/4
    exif:FocalLengthIn35mmFilm: 26
    exif:GPSAltitude: 1300306/4615
    exif:GPSAltitudeRef: 0
    exif:GPSDestBearing: 861598/3733
    exif:GPSDestBearingRef: T
    exif:GPSHPositioningError: 93285/11248
    exif:GPSImgDirection: 861598/3733
    exif:GPSImgDirectionRef: T
    exif:GPSInfo: 2264
    exif:GPSLatitude: 29/1, 36/1, 720/100
    exif:GPSLatitudeRef: N
    exif:GPSLongitude: 98/1, 26/1, 5815/100
    exif:GPSLongitudeRef: W
    exif:GPSSpeed: 63415/1154601
    exif:GPSSpeedRef: K
    exif:ImageDescription: Benny.
    exif:LensMake: Apple
    exif:LensModel: iPhone 11 back dual wide camera 4.25mm f/1.8
    exif:LensSpecification: 807365/524263, 17/4, 9/5, 12/5
    exif:Make: Apple
    exif:MakerNote: 65, 112, 112, 108, 101, 32, 105, 79, 83, 0, 0, 1, 77, 77, 0, 37, 0, 1, 0, 9, 0, 0, 0, 1, 0, 0, 0, 12, 0, 2, 0, 7, 0, 0, 2, 46, 0, 0, 1, 208, 0, 3, 0, 7, 0, 0, 0, 104, 0, 0, 3, 254, 0, 4, 0, 9, 0, 0, 0, 1, 0, 0, 0, 1, 0, 5, 0, 9, 0, 0, 0, 1, 0, 0, 0, 94, 0, 6, 0, 9, 0, 0, 0, 1, 0, 0, 0, 102, 0, 7, 0, 9, 0, 0, 0, 1, 0, 0, 0, 1, 0, 8, 0, 10, 0, 0, 0, 3, 0, 0, 4, 102, 0, 12, 0, 10, 0, 0, 0, 2, 0, 0, 4, 126, 0, 13, 0, 9, 0, 0, 0, 1, 0, 0, 0, 23, 0, 14, 0, 9, 0, 0, 0, 1, 0, 0, 0, 0, 0, 15, 0, 9, 0, 0, 0, 1, 0, 0, 0, 2, 0, 16, 0, 9, 0, 0, 0, 1, 0, 0, 0, 1, 0, 20, 0, 9, 0, 0, 0, 1, 0, 0, 0, 11, 0, 23, 0, 9, 0, 0, 0, 1, 38, 0, 0, 0, 0, 25, 0, 9, 0, 0, 0, 1, 0, 0, 0, 0, 0, 26, 0, 2, 0, 0, 0, 6, 0, 0, 4, 142, 0, 31, 0, 9, 0, 0, 0, 1, 0, 0, 0, 0, 0, 32, 0, 2, 0, 0, 0, 37, 0, 0, 4, 148, 0, 33, 0, 10, 0, 0, 0, 1, 0, 0, 4, 186, 0, 35, 0, 9, 0, 0, 0, 2, 0, 0, 4, 194, 0, 37, 0, 9, 0, 0, 0, 1, 0, 0, 28, 12, 0, 38, 0, 9, 0, 0, 0, 1, 0, 0, 0, 3, 0, 39, 0, 10, 0, 0, 0, 1, 0, 0, 4, 202, 0, 40, 0, 9, 0, 0, 0, 1, 0, 0, 0, 1, 0, 43, 0, 2, 0, 0, 0, 37, 0, 0, 4, 210, 0, 45, 0, 9, 0, 0, 0, 1, 0, 0, 12, 64, 0, 46, 0, 9, 0, 0, 0, 1, 0, 0, 0, 1, 0, 47, 0, 9, 0, 0, 0, 1, 0, 0, 0, 91, 0, 51, 0, 9, 0, 0, 0, 1, 0, 0, 32, 0, 0, 52, 0, 9, 0, 0, 0, 1, 0, 0, 0, 10, 0, 53, 0, 9, 0, 0, 0, 1, 0, 0, 0, 9, 0, 54, 0, 9, 0, 0, 0, 1, 0, 0, 0, 7, 0, 55, 0, 9, 0, 0, 0, 1, 0, 0, 0, 4, 0, 58, 0, 9, 0, 0, 0, 1, 0, 0, 0, 4, 0, 59, 0, 9, 0, 0, 0, 1, 0, 0, 0, 0, 0, 60, 0, 9, 0, 0, 0, 1, 0, 0, 0, 4, 0, 0, 0, 0, 98, 112, 108, 105, 115, 116, 48, 48, 79, 17, 2, 0, 137, 1, 232, 0, 228, 0, 246, 0, 14, 1, 17, 1, 14, 1, 8, 1, 210, 0, 177, 0, 131, 0, 116, 0, 101, 0, 113, 0, 106, 0, 72, 0, 145, 1, 210, 0, 204, 0, 245, 0, 1, 1, 3, 1, 246, 0, 218, 0, 172, 0, 157, 0, 119, 0, 117, 0, 105, 0, 122, 0, 98, 0, 107, 0, 252, 1, 243, 0, 204, 0, 167, 0, 146, 0, 210, 0, 222, 0, 188, 0, 162, 0, 138, 0, 121, 0, 137, 0, 125, 0, 131, 0, 98, 0, 104, 0, 254, 0, 252, 0, 182, 0, 40, 0, 30, 0, 108, 0, 165, 0, 172, 0, 150, 0, 127, 0, 116, 0, 129, 0, 128, 0, 117, 0, 97, 0, 103, 0, 202, 0, 224, 0, 201, 0, 28, 0, 31, 0, 34, 0, 53, 0, 92, 0, 58, 0, 63, 0, 104, 0, 116, 0, 140, 0, 105, 0, 96, 0, 112, 0, 203, 0, 206, 0, 138, 0, 36, 0, 31, 0, 36, 0, 45, 0, 50, 0, 28, 0, 29, 0, 39, 0, 90, 0, 124, 0, 95, 0, 95, 0, 114, 0, 185, 0, 187, 0, 57, 0, 36, 0, 43, 0, 67, 0, 49, 0, 67, 0, 23, 0, 18, 0, 23, 0, 90, 0, 134, 0, 88, 0, 94, 0, 109, 0, 198, 0, 186, 0, 50, 0, 30, 0, 23, 0, 72, 0, 48, 0, 53, 0, 75, 0, 33, 0, 32, 0, 154, 0, 69, 0, 79, 0, 92, 0, 103, 0, 155, 0, 152, 0, 56, 0, 32, 0, 17, 0, 45, 0, 24, 0, 41, 0, 58, 0, 62, 0, 116, 0, 161, 0, 63, 0, 71, 0, 95, 0, 103, 0, 172, 0, 108, 0, 50, 0, 30, 0, 26, 0, 35, 0, 29, 0, 29, 0, 49, 0, 44, 0, 166, 0, 125, 0, 22, 0, 154, 0, 123, 0, 97, 0, 91, 0, 51, 0, 48, 0, 35, 0, 33, 0, 40, 0, 42, 0, 32, 0, 63, 0, 38, 0, 104, 0, 102, 0, 114, 0, 116, 0, 130, 0, 108, 0, 47, 0, 90, 0, 56, 0, 37, 0, 18, 0, 27, 0, 37, 0, 31, 0, 29, 0, 20, 0, 134, 0, 192, 0, 152, 0, 141, 0, 162, 0, 137, 0, 57, 0, 124, 0, 85, 0, 44, 0, 30, 0, 47, 0, 40, 0, 64, 0, 49, 0, 63, 0, 179, 0, 214, 0, 192, 0, 162, 0, 168, 0, 175, 0, 87, 0, 143, 0, 135, 0, 43, 0, 32, 0, 26, 0, 77, 0, 211, 0, 197, 0, 211, 0, 201, 0, 221, 0, 191, 0, 173, 0, 184, 0, 183, 0, 87, 0, 130, 0, 141, 0, 50, 0, 27, 0, 58, 0, 145, 0, 217, 0, 223, 0, 220, 0, 222, 0, 227, 0, 203, 0, 182, 0, 188, 0, 185, 0, 76, 0, 147, 0, 154, 0, 118, 0, 68, 0, 115, 0, 237, 0, 233, 0, 218, 0, 234, 0, 231, 0, 225, 0, 202, 0, 189, 0, 200, 0, 184, 0, 0, 8, 0, 0, 0, 0, 0, 0, 2, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 12, 98, 112, 108, 105, 115, 116, 48, 48, 212, 1, 2, 3, 4, 5, 6, 7, 8, 85, 102, 108, 97, 103, 115, 85, 118, 97, 108, 117, 101, 89, 116, 105, 109, 101, 115, 99, 97, 108, 101, 85, 101, 112, 111, 99, 104, 16, 1, 19, 0, 2, 203, 8, 14, 89, 117, 237, 18, 59, 154, 202, 0, 16, 0, 8, 17, 23, 29, 39, 45, 47, 56, 61, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 9, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 63, 255, 255, 239, 147, 0, 5, 214, 77, 255, 255, 169, 183, 0, 0, 127, 67, 255, 255, 212, 145, 0, 0, 59, 201, 0, 0, 0, 73, 0, 0, 1, 5
    exif:MeteringMode: 5
    exif:Model: iPhone 11
    exif:OffsetTime: -05:00
    exif:OffsetTimeDigitized: -05:00
    exif:OffsetTimeOriginal: -05:00
    exif:PhotographicSensitivity: 1600
    exif:PixelXDimension: 4032
    exif:PixelYDimension: 3024
    exif:SceneCaptureType: 0
    exif:SceneType: 1
    exif:SensingMethod: 2
    exif:ShutterSpeedValue: 74706/19121
    exif:Software: 14.7.1
    exif:SubjectArea: 2009, 1509, 2318, 1390
    exif:SubSecTimeDigitized: 271
    exif:SubSecTimeOriginal: 271
    exif:TileLength: 512
    exif:TileWidth: 512
    exif:WhiteBalance: 0
    icc:copyright: Copyright Apple Inc., 2017
    icc:description: Display P3
    signature: daa25d88b66f1248b6f382b59840e871796490212dc5f05430f29aba88f36136
    unknown: iPhone 11
  Tainted: False
  Filesize: 0B
  Number pixels: 12.1928M
  Pixels per second: 6.37543GP
  User time: 0.016u
  Elapsed time: 0:01.001
  Version: ImageMagick 7.1.0-4 Q16 x64 2021-07-18 https://imagemagick.org
```
Note that for me (IM v7.1.0-4), IM says the image has RGB channels. I suspect for you that it says something different. If so, that would suggest a recent bug in IM.

Note also it says there is an embedded "Display P3" profile.

**Comment by davidsalazar** [comment, marked as answer]
Thanks to everyone, I'm writing this as a separate comment so I can mark this as an answer so nobody makes this dumb mistake I did. 

Everyone, if you're converting an iphone heic to srgb then stripping all exif, please use the srgb provided in the imagemagick package. srgb.icc don't make the mistake I was doing and using a diff one. 

Final cmd will look something like this.
`magick IMG_2253.HEIC -profile srgb.icc -strip x1.jpg`
